### PR TITLE
Add interface to `astra` toolbox cone beam projectors

### PR DESCRIPTION
**Pull request by bwohlberg**
Add interface to `astra` toolbox cone beam X-ray projectors.

The initial version of this module was constructed from the parallel beam interface with the assistance of Claude 4.5 Sonnet.

### CODE REVIEW

File: scico/linop/xray/astra_cone.py [hunk 1]
```diff
@@ -0,0 +1,413 @@
+# -*- coding: utf-8 -*-
+# Copyright (C) 2026 by SCICO Developers
+# All rights reserved. BSD 3-clause License.
+# This file is part of the SCICO package. Details of the copyright and
+# user license can be found in the 'LICENSE' file distributed with the
+# package.
+
+r"""X-ray cone beam transform LinearOperators wrapping the ASTRA toolbox.
```
Comment (Michael-T-McCann) [inline scico/linop/xray/astra_cone.py]:
```suggestion
r"""X-ray cone beam transform :class:`.LinearOperator`s wrapping the ASTRA toolbox.
```

Comment (Michael-T-McCann) [inline scico/linop/xray/astra_cone.py]: Suggestion: "LinearOperator(s?) wrapping the cone beam X-ray transform from the ASTRA Toolbox".

File: scico/linop/xray/astra_cone.py [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+
+X-ray cone beam transform :class:`.LinearOperator` wrapping the cone beam
+projections in the
+`ASTRA toolbox <https://github.com/astra-toolbox/astra-toolbox>`_.
```
Comment (Michael-T-McCann) [inline scico/linop/xray/astra_cone.py]: Sorry if pedantic, but I found this tough to read. I think you can say "LinearOperator wrapping the cone beam projectors in ASTRA" or "Cone beam projection LinearOperator implemented using ASTRA", but not "Cone beam projection LinearOperator wrapping the cone beam projectors in ASTRA"

File: scico/linop/xray/astra_cone.py [hunk 3]
```diff
@@ -0,0 +1,413 @@
+# -*- coding: utf-8 -*-
+# Copyright (C) 2026 by SCICO Developers
+# All rights reserved. BSD 3-clause License.
+# This file is part of the SCICO package. Details of the copyright and
+# user license can be found in the 'LICENSE' file distributed with the
+# package.
+
+r"""X-ray cone beam transform LinearOperators wrapping the ASTRA toolbox.
+
+X-ray cone beam transform :class:`.LinearOperator` wrapping the cone beam
```
Comment (Michael-T-McCann) [inline scico/linop/xray/astra_cone.py]: "X-ray cone beam transform" has 6 results on Google, while "cone beam X-ray transform" has 937, the latter sounds more correct to me.

File: scico/linop/xray/astra_cone.py [hunk 4]
```diff
@@ -0,0 +1,413 @@
...
+projections in the
+`ASTRA toolbox <https://github.com/astra-toolbox/astra-toolbox>`_.
+This package provides CUDA implementations of core functionality for cone
+beam geometries.
```
Comment (Michael-T-McCann) [inline scico/linop/xray/astra_cone.py]: Seems out of place.

File: scico/linop/xray/astra_cone.py [hunk 5]
```diff
@@ -0,0 +1,413 @@
...
+    """Ensure that `x.flags.writeable` is ``True``, copying if needed."""
+    if hasattr(x, "flags"):  # x is a numpy array
+        if not x.flags.writeable:
+            try:
+                x.setflags(write=True)
+            except ValueError:
+                x = x.copy()
+    else:  # x is a jax array (which is immutable)
+        x = np.array(x)
+    return x
```
Comment (Michael-T-McCann) [inline scico/linop/xray/astra_cone.py:56]: If this is copied from the other astra wrapper, consider moving to a shared util file.

File: scico/linop/xray/astra_cone.py [hunk 6]
```diff
@@ -0,0 +1,413 @@
...
+
+    In the "cone" case, the source and detector rotate clockwise about the
+    `z` axis in the `x`-`y` plane. The source is positioned at distance
+    `source_dist` from the origin, and the detector is at distance
+    `det_dist` from the origin (making the source-detector distance
+    `source_dist + det_dist`).
+
+    In the "cone_vec" case, each view is determined by vectors specifying:
+    the source position :math:`\mb{s}`, detector center :math:`\mb{d}`,
+    and detector basis vectors :math:`\mb{u}` and :math:`\mb{v}`.
```
Comment (Michael-T-McCann) [inline scico/linop/xray/astra_cone.py]: Duplicated from above, probably makes more sense to leave it here.

File: scico/test/linop/xray/test_astra_cone.py [hunk 7]
```diff
@@ -0,0 +1,461 @@
...
+
+@pytest.mark.skipif(jax.devices()[0].platform != "gpu", reason="GPU required for cone beam")
+def test_ATA_call(testobj):
+    """Test x A^T A x = || A x ||_2^2 property using call interface."""
+    Ax = testobj.A(testobj.x)
+    ATAx = testobj.A.adj(Ax)
+    n0 = np.sum(testobj.x * ATAx)
+    n1 = np.linalg.norm(Ax) ** 2
+    assert np.abs(n0 - n1) / max(n0, n1) < 0.5  # poorly matched adjoint
+
```
Comment (Michael-T-McCann) [inline scico/test/linop/xray/test_astra_cone.py:185]: Don't we have a function to test adjoints?

File: scico/test/linop/xray/test_astra_cone.py [hunk 8]
```diff
@@ -0,0 +1,461 @@
...
+@pytest.mark.skipif(jax.devices()[0].platform != "gpu", reason="GPU required for cone beam")
+def test_AAT_matmul(testobj):
+    """Test y A A^T y = || A^T y ||_2^2 property using matmul interface."""
+    ATy = testobj.A.T @ testobj.y
+    AATy = testobj.A @ ATy
+    n0 = np.sum(testobj.y * AATy)
+    n1 = np.linalg.norm(ATy) ** 2
+    assert np.abs(n0 - n1) / max(n0, n1) < 0.75  # poorly matched adjoint
+
```
Comment (Michael-T-McCann) [inline scico/test/linop/xray/test_astra_cone.py:215]: These seem to be testing LinearOperator more than they test XRayTransform3DCone

File: scico/test/linop/xray/test_astra_cone.py [hunk 9]
```diff
@@ -0,0 +1,461 @@
...
+        angles=np.array([0.0]),
+        source_dist=200.0,
+        det_dist=200.0,
+    )
+
+    y_close = A_close @ x
+    y_far = A_far @ x
+
+    # Closer source should produce much larger projection
+    assert y_close.sum() > 2 * y_far.sum()
```
Comment (Michael-T-McCann) [inline scico/test/linop/xray/test_astra_cone.py]: "magnification" suggests we expect a projection with a larger physical width (more pixels on the detector. But this comparing the sums of the two projections. It's not obvious to me how these should compare, it depends on how the operator is normalized.

File: scico/test/linop/xray/test_astra_cone.py [hunk 10]
```diff
@@ -0,0 +1,461 @@
...
+    assert rel_res(x, np.clip(x_fdk, 0.0, 1.0)) < 0.3
+
+
+def test_ensure_writeable():
+    """Test the _ensure_writeable utility function."""
+    # Test with numpy array
+    assert isinstance(_ensure_writeable(np.ones((2, 1))), np.ndarray)
+
+    # Test with JAX array
+    assert isinstance(_ensure_writeable(snp.ones((2, 1))), np.ndarray)
```
Comment (Michael-T-McCann) [inline scico/test/linop/xray/test_astra_cone.py:461]: If this function moves, the test should, too.

**Review by Michael-T-McCann — CHANGES_REQUESTED** [review]

File: scico/linop/xray/astra_cone.py (hunk 1 above)
Comment (bwohlberg) [reply scico/linop/xray/astra_cone.py]: Adding the :class: formatting pushes the initial line of the docstring past 72 characters; while it would be nice to include the formatting, it would contravene our (and standard Python) coding convention. There's no perfect solution, but I'm inclined to leave it as-is, particularly since the link is in the longer text directly below.

Comment (bwohlberg) [reply scico/linop/xray/astra_cone.py]: Agreed and done.

File: scico/linop/xray/astra_cone.py (hunk 2 above)
Comment (bwohlberg) [reply scico/linop/xray/astra_cone.py]: Fixed.

File: scico/linop/xray/astra_cone.py (hunk 4 above)
Comment (bwohlberg) [reply scico/linop/xray/astra_cone.py]: Corresponding text is in the `astra` module. I made improvements to both. Feel free to edit if not satisfied.

File: scico/linop/xray/astra_cone.py (hunk 5 above)
Comment (bwohlberg) [reply scico/linop/xray/astra_cone.py:56]: Changes in another pending PR will remove this function, so there isn't any point to do it here.

File: scico/linop/xray/astra_cone.py (hunk 6 above)
Comment (bwohlberg) [reply scico/linop/xray/astra_cone.py]: Fixed

File: scico/test/linop/xray/test_astra_cone.py (hunk 7 above)
Comment (bwohlberg) [reply scico/test/linop/xray/test_astra_cone.py:185]: Yes. This was copied from the parallel beam tests. If a serious concern, I suggest a separate PR to modify both.

File: scico/test/linop/xray/test_astra_cone.py (hunk 8 above)
Comment (bwohlberg) [reply scico/test/linop/xray/test_astra_cone.py:215]: As above; agreed, but the flaw is already in the parallel beam interface.

File: scico/test/linop/xray/test_astra_cone.py (hunk 9 above)
Comment (bwohlberg) [reply scico/test/linop/xray/test_astra_cone.py]: Would you be happier with comparing the sizes of the supports?

Comment (Michael-T-McCann) [reply scico/test/linop/xray/test_astra_cone.py]: Yes

**Review by Michael-T-McCann — APPROVED** [review]
